Migrate watcher to system indices infrastructure

pugnascotia opened the pull request:
Part of #61656.

Migrate the `.watches` and `.triggered_watches` system indices to use the auto-create infrastructure. The watcher history indices are left alone.

As part of this work, a `SystemIndexDescriptor` now inspects its mappings to determine whether it has any dynamic mappings. This influences how strict Elasticsearch is with enforcing the descriptor's mappings, since ES cannot know in advanced what all the mappings will be.

This PR also fixes the `SystemIndexManager` so that it doesn't fall over when attempting to inspect the state of an index that hasn't been created yet.

pugnascotia committed:
WIP - migrate watcher to system indices infrastructure

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

jaymode commented on x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java:
```diff
@@ -602,9 +719,14 @@ static void validAutoCreateIndex(Settings settings, Logger logger) {
-        String errorMessage = LoggerMessageFormat.format("the [action.auto_create_index] setting value [{}] is too" +
-                " restrictive. disable [action.auto_create_index] or set it to " +
-                "[{},{}]", (Object) value, Watch.INDEX, TriggeredWatchStoreField.INDEX_NAME);
+        String errorMessage = LoggerMessageFormat.format(
```
We should not need this anymore since these are system indices and we should always be able to create the watches index and triggered watches index.

pugnascotia replied:
I've removed this code.

pugnascotia committed:
Remove redundant action.auto_create_index check

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

jaymode reviewed (approved):
LGTM

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

mark-vieira commented:
@elasticmachine update branch

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

pugnascotia commented:
There's something very peculiar going on here. As far as I can tell:

1. The Watcher asciidoc tests create a new watch. In the request payload, the new watch populates a `metadata` key with `{ color: red }`, which is legitimate.
2. Watcher submits an index request for the new watch
3. This request gets turned into a bulk request.
4. The bulk request notices that the `.watches` index doesn't exist, and submits an `AutoCreateAction` to create it.
5. The index is created, as far as I can tell with the correct mappings from the system index descriptor
6. But then a `TransportAutoPutMappingAction` fails because it tries to apply some incorrect mappings to the `.watches` system index - and it so happens that these new mappings match the `metadata` from the create watch payload.
7. As far as I can see, this action is raised as part of the replication code, at which point I'm lost.

jaymode commented:
> But then a TransportAutoPutMappingAction fails because it tries to apply some incorrect mappings to the .watches system index - and it so happens that these new mappings match the metadata from the create watch payload. There's no reason that I can see why this metadata from the watcher payload should have any bearing on the mappings of the .watches index (but I could be wrong)

I think the issue is that watches have dynamic objects within the mapping including metadata among others. So when the watch is stored with metadata then the mappings need to be updated to account for the `metadata.color` field. I think Security will have the same issue for dynamic fields.

In order to resolve this, I think a change is needed to how mappings are validated. If a mapping contains dynamic fields then any fields contained within that object should be allowed. Make sense?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

pugnascotia committed:
Check for dynamic fields in a SystemIndexDescriptor's mappings

pugnascotia committed:
Merge snafu

pugnascotia committed:
Revert whitespace changes

pugnascotia committed:
Remove references to deleted templates

pugnascotia committed:
Remove redundant tests

pugnascotia committed:
Test fix

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher

pugnascotia committed:
Handle non-existent indices in SystemIndexManager

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 61656-auto-create-watcher

martijnvg commented on x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java:
```diff
@@ -710,4 +673,166 @@ public String getFeatureName() {
+                builder.field("dynamic", "strict");
+                {
+                    builder.startObject("_meta");
+                    builder.field("version", Version.CURRENT);
```
The current template doesn't have a `version` field under they `_meta` object.
If I read the code correctly in `SystemIndexManager` this means that irregardless of that, a mapping update will be triggered?

pugnascotia replied:
@martijnvg it turns out that the `SystemIndexManager` would never have updated the mappings, because without a value for `_metadata.version`, the manager would always think that the version was the latest (because `Version.fromString(null)` returns `Version.CURRENT`). I've pushed a fix for that case.

martijnvg replied:
Thanks for fixing that!

martijnvg reviewed (approved):
From Watcher side LGTM. I did left a question for my own understanding.

pugnascotia committed:
Tweak null metadata version handling in SystemIndexManager

`SystemIndexManager.readMappingVersion` didn't take into account that
calling `Version.readString(null)` would return `Version.CURRENT`, which
would mean that an index with version metadata would never have its
mappings updated. Change the logic to return `Version.EMPTY` in this
case, so that the mappings will be updated.

pugnascotia committed:
Merge commit '0daae1ec2de' into 61656-auto-create-watcher

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-watcher
